### PR TITLE
Accelerometer implementation for CoreMotion Framework and correspondi…

### PR DESCRIPTION
**Pull request by afaruqui**
…ng addtion to WOCCatalog sample

### CODE REVIEW

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 1]
```diff
@@ -15,34 +16,124 @@
...
+       if (!checkDefaultAccelerometer()) {                
+            _accelerometerAvailable = false;
+            _accelerometerActive = false;
+            _accelerometerUpdateInterval = 0;
+        } else {
+            _accelerometer = [WDSAccelerometer getDefault];
+            _accelerometerAvailable = true;
+            _accelerometerActive = false;
+            _accelerometerUpdateInterval = _accelerometer.minimumReportInterval;
+            _accelerometer.reportInterval = _accelerometer.minimumReportInterval;
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: Won't this turn the Accelerometer on prematurely?

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 2]
```diff
@@ -15,34 +16,124 @@
...
+@implementation CMMotionManager 
+
+- (instancetype)init {
+    if (self = [super init]) {        
+        
+        // Calling into C++/CX to determine if the default accelerometer device is available
+        // This step is necessary for now as this call using projections [WDSAccelerometer getDefault] does not return NULL
+        // as expected for the case where no accelerometer device is available
+       if (!checkDefaultAccelerometer()) {                
+            _accelerometerAvailable = false;
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: Should we add a call to NSTraceInfo() here so it's more apparent to developers that we failed to init?

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 3]
```diff
@@ -15,34 +16,124 @@
...
+
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 - (void)startAccelerometerUpdates {
-    UNIMPLEMENTED();
+    
+    if (self.accelerometerAvailable) {
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: Any threading concerns on any of these accessors/methods?  Should we synchronize on self?

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 4]
```diff
@@ -15,34 +16,124 @@
...
 */
-- (void)stopAccelerometerUpdates {
-    UNIMPLEMENTED();
-}
+- (void) stopAccelerometerUpdates {
+
+    if (self.accelerometerActive) {
+        self.accelerometerActive = false;
+        self.accelerometer.reportInterval = 0;
+        [self.accelerometer removeReadingChangedEvent:self.accelerometerToken];
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: should clear out self.accelerometerToken too, and should check that it's actually set before trying to unsubscribe.

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 5]
```diff
@@ -15,34 +16,124 @@
...
+    
+    if (self.accelerometerAvailable) {
+        
+        if (self.accelerometerActive) {
+            [self stopAccelerometerUpdates];
+        }
+        
+        self.accelerometerActive = true;
+        
+        if (self.accelerometerUpdateInterval < self.accelerometer.minimumReportInterval) {
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: nit: should we NSTraceInfo here if we're capping the min interval?

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 6]
```diff
@@ -15,34 +16,124 @@
...
  @Notes
 */
-- (void)stopAccelerometerUpdates {
-    UNIMPLEMENTED();
-}
+- (void) stopAccelerometerUpdates {
+
+    if (self.accelerometerActive) {
+        self.accelerometerActive = false;
+        self.accelerometer.reportInterval = 0;
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: nit: a comment here may be nice, as setting to 0 is what 'stops' the accelerometer.

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 7]
```diff
@@ -15,34 +16,124 @@
...
 */
-- (void)startAccelerometerUpdatesToQueue:(NSOperationQueue*)queue withHandler:(CMAccelerometerHandler)handler {
-    UNIMPLEMENTED();
+- (void)startAccelerometerUpdatesToQueue:(NSOperationQueue*)queue 
+                                           withHandler:(CMAccelerometerHandler)handler {
+    if (self.accelerometerAvailable) {
+        [self startAccelerometerUpdates];
+        self.accelerometerQueue = queue;
+
+        self.accelerometerToken = [self.accelerometer addReadingChangedEvent:
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: What if someone calls this multiple times on the same motionManager? is that valid?

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 8]
```diff
@@ -15,34 +16,124 @@
 //******************************************************************************
 
 #import <CoreMotion/CMMotionManager.h>
+#import <CoreMotion/CMAccelerometerData.h>
 #import <StubReturn.h>
+#import "CMAccelerometerDataInternal.h"
+#import "CMMotionManagerWinRT.h"
+
+typedef wchar_t WCHAR;
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: This should no longer be necessary

**Comment by afaruqui** [comment]
Thank You for the feedback. We addressed the comments and made appropriate changes.

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 9]
```diff
@@ -15,35 +16,149 @@
...
+    WDSAccelerometerReading* reading = [self.accelerometer getCurrentReading];
+
+    _accelerometerData = [[CMAccelerometerData alloc] initWithValues:reading.accelerationX 
+                                                                   y:reading.accelerationY
+                                                                   z:reading.accelerationZ];
+    return _accelerometerData;
+ }
+
+
+-(void)setAccelerometerReportInterval {
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: nit: seems a bit overkill to pull this out as a helper method.

Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: private methods should start with '_'

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 10]
```diff
@@ -15,35 +16,149 @@
...
+    self.accelerometer.reportInterval = static_cast<unsigned int>(self.accelerometerUpdateInterval * 1000);
 }
 
+
+
+-(void)setAccelerometerUpdateInterval:(NSTimeInterval)updateInterval {
+    @synchronized(self) {
+        
+        if (updateInterval * 1000 < self.accelerometer.minimumReportInterval) {
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: Can you add comments for the conversion here? Something simple like 'IOS uses seconds whereas WinRT uses milliseconds' would be helpful.  Bonus points for pulling 1000 out as a constant that is descriptively named.

File: Frameworks/CoreMotion/CMMotionManager.mm [hunk 11]
```diff
@@ -15,35 +16,149 @@
...
+
+
+-(void)setAccelerometerReportInterval {
+    self.accelerometer.reportInterval = static_cast<unsigned int>(self.accelerometerUpdateInterval * 1000);
 }
 
+
+
+-(void)setAccelerometerUpdateInterval:(NSTimeInterval)updateInterval {
+    @synchronized(self) {
```
Comment (jaredhms) [inline Frameworks/CoreMotion/CMMotionManager.mm]: why are we locking here, but nowhere else?

**Comment by afaruqui** [comment]
Thank you for the feedback. We addressed the lastest comments and uploaded the changes.

**Pull request merged**